### ISSUE TITLE
Inefficient code generated for `__builtin_convertvector` to a boolean vector on ARM.

### ISSUE THREAD

**Issue by lawben** [llvm/llvm-project#59829]
While playing around with converting vector intrinsics comparison results to masks, I noticed that the generated code for ARM is rather inefficient compared to how it could be. 

Link to godbolt for code: [https://godbolt.org/z/3jdd9b76d](https://godbolt.org/z/3jdd9b76d)

Considering the following code that compares to vectors and creates a mask from this, i.e., a boolean vector where each value is represented by one bit instead of 16 in this example.

```cpp
using VecU16x8 __attribute__((vector_size(16), aligned(16))) = uint16_t;
using MaskT __attribute__((ext_vector_type(8))) = bool;

uint8_t mask_from_builtin_all_ones_or_zeros(VecU16x8 x, VecU16x8 y) {
    auto matches = x == y;
    auto result = __builtin_convertvector(matches, MaskT);
    return reinterpret_cast<uint8_t&>(result);
}
```

This generates quite a lot of instructions and has worse performance than a more commonly used approach on ARM with NEON, which uses a mask and "horizontally" adds all values in a vector.

```cpp
uint8_t mask_manual_all_ones_or_zeros(VecU16x8 a, VecU16x8 b) {
    auto matches = vceqq_u16(a, b);
    constexpr VecU16x8 mask = {1, 2, 4, 8, 16, 32, 64, 128};
    return vaddvq_u16(vandq_u16(matches, mask));
}
```

This can also be applied to `uint8/32/64_t`  with similar masks/instructions.

This mask "trick" only works if the compiler can ensure that all bits are either 1 or 0, which it can after a vector comparison. To demonstrate this, the [godbolt example](https://godbolt.org/z/3jdd9b76d) also contains a x86 version, in which in clang generates different instructions based on this knowledge. 

A tiny benchmark on my M1 Pro MacBook shows that the `mask_manual_all_ones_or_zeros` code runs ~4x faster than the version generated by `__builtin_convertvector`. So while the NEON version is not always possible (unless the x86 path is chosen and you compare the vector to itself to ensure equality sets all bits to 1 or 0), it could be applied to improve performance on ARM in a common case.

**Comment by lawben** [comment]
Just to document this in case anyone is looking into this or wants to: I've been working on a patch for this. I have an initial first implementation that generates the more efficient version on AArch64. The code still needs to be cleaned up and tested, but conceptually it works.

**Comment by lawben** [comment]
Closed via [3653722](https://github.com/llvm/llvm-project/commit/3653722ce6d9bb44a42ffdce4e07080aa1928ddc).
